### PR TITLE
Add secure memcmp and memzero implementation.

### LINKED ISSUE #347: Implement our own secure `memcmp` and `memzero` if libsodium isn't available
See `crypto_core.c`, the implementations of `crypto_memcmp` and `crypto_memzero`.

### PR DESCRIPTION
**Pull request by Diadlo**
Implementation mostly taken from libsodium utils.

Fix #347

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/354)

### CODE REVIEW

**Comment by nurupo** [comment]
Just for reference, here is the relevant code section from libsodium https://github.com/jedisct1/libsodium/blob/9cae7b6b7c6f527df807157602ba48821281048b/src/libsodium/sodium/utils.c#L61-L132.

**Comment by nurupo** [comment]
@Diadlo any reason why you didn't include the copyright and the license statements from libsodium? You are violating libsodium's licensing terms by copy-pasting the code like that.

**Comment by Diadlo** [comment]
@nurupo Sorry, I'm don't understand all "copyright and the license" stuff. Where I should paste copyright?
I should paste full licence text? Or just header?
```
/*
 * ISC License
 *
 * Copyright (c) 2013-2016
 * Frank Denis <j at pureftpd dot org>
 */
```

**Comment by nurupo** [comment]
libsodium is licensed under ISC license https://github.com/jedisct1/libsodium/blob/master/LICENSE.

First off, if we want to include libsodium's code in toxcore, we need to make sure ISC is GPL-compatible, which is the case https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses.

Now, we need to comply with the terms on the ISC license

> * Permission to use, copy, modify, and/or distribute this software for any
> * purpose with or without fee is hereby granted, provided that the above
> * copyright notice and this permission notice appear in all copies.

So we need to include the copyright notice and the license statement. Not sue if saying that the original functions were licensed under ISC license is enough or we should include the whole license text though. I think something like:

>crypto_memcmp() and crypto_memzero() functions are taken from libsodium
> Copyright (c) 2013-2016 Frank Denis <j at pureftpd dot org>, ISC License

would be compliant with the ISC license. The first line telling which functions were taken from where shouldn't be required, but I like to have it for clarity.

@iphydf what do you thing?

**Comment by nurupo** [comment]
@iphydf suggested placing the borrowed functions into a separate file, which would have the entire https://github.com/jedisct1/libsodium/blob/master/LICENSE as its header text. I like that idea.

**Comment by Diadlo** [comment]
Done

**Comment by nurupo** [comment]
Where are `HAVE_EXPLICIT_BZERO` and `HAVE_MEMSET_S` defined? I believe libsodium's autotools build system defines those.

**Comment by Diadlo** [comment]
@nurupo Removed

**Comment by nurupo** [comment]
Not sure if removing them in favor of the generic solution is a good idea.

**Comment by Diadlo** [comment]
@nurupo I found definition only [`HAVE_EXPLICIT_BZERO`](https://github.com/jedisct1/libsodium/blob/master/configure.ac#L678). I will create quiestion-issue to libsodium

**Comment by Diadlo** [comment]
@nurupo Done.

**Comment by Diadlo** [comment]
Fixed

**Comment by nurupo** [comment]
Don't know whether the `CRYPTO_TEST_MEMCMP_EPS` constant is small enough, it might very well be, but other than that the tests look good.

**Comment by nurupo** [comment]
Also, I don' think we need `extern "C" {` block in headers that are not public.

**Comment by Diadlo** [comment]
Without `extern "C" (` windows build was [failed](https://travis-ci.org/TokTok/c-toxcore/jobs/186697508#L1048)

`diff_same` ~ `diff_not_same` ~ 4250 on my PC, which is 10 times greater than `EPS`
Should I run many tests and calculate expected value for the current machine and variance?

**Comment by nurupo** [comment]
>Without extern "C" ( windows build was failed

That's because the names collide. libsodium defines `sodium_memzero()` and we define it too. Having different linkage, C vs C++, removes this collision.

>diff_same ~ diff_not_same ~ 4250 on my PC, which is 10 times greater than EPS
Should I run many tests and calculate expected value for the current machine and variance?

I'm not knowledgeable in the proper way of checking if a function is constant time or not. The test looks good to me.

File: CMakeLists.txt.user [hunk 1]
```diff
@@ -0,0 +1,371 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (sudden6) [inline CMakeLists.txt.user]: you might want to add `.user` to your gitignore

**Comment by Diadlo** [comment]
Sometimes `test_memcmp` fails (only for clang travis) with delta time `1000` or `2000`. I haven't any idea

**Comment by iphydf** [comment]
Actually, one more thing: `sodium_utils` is not a great name. Move the definition of `crypto_mem*` into the separate file, so you get something like:
```c
void crypto_memzero(void *data, size_t length)
{
#ifndef VANILLA_NACL
    sodium_memzero(data, length);
#else
    // body of current sodium_memzero here
#endif
}
```
and the same for `crypto_memcmp`.

**Comment by nurupo** [comment]
@iphydf Yep, doing so will also fix the conflicting declaration issue mentioned in https://github.com/TokTok/c-toxcore/pull/354#issuecomment-269165839, as we won't be declaring our own `sodium_memzero()` and `sodium_memcpm()`, so we can drop that dirty `extern "C" (` hack.

**Comment by zetok** [comment]
Wouldn't it be better to just remove the support for NaCl and use only libsodium?

File: CMakeLists.txt [hunk 2]
```diff
@@ -204,7 +204,11 @@ apidsl(
   toxcore/crypto_core.api.h)
 add_module(toxcrypto
   toxcore/crypto_core.c
-  toxcore/crypto_core.h)
+  toxcore/crypto_core.h
+  toxcore/crypto_core_mem.c)
+include(CheckFunctionExists)
+CHECK_FUNCTION_EXISTS(explicit_bzero HAVE_EXPLICIT_BZERO)
+CHECK_FUNCTION_EXISTS(memset_s HAVE_MEMSET_S)
```
Comment (iphydf) [inline CMakeLists.txt]: Lowercase function calls: `check_function_exists`.

File: auto_tests/crypto_test.c [hunk 3]
```diff
@@ -3,6 +3,7 @@
 #endif
 
 #include "../toxcore/net_crypto.h"
+#include "../toxcore/misc_tools.h"
```
Comment (iphydf) [inline auto_tests/crypto_test.c]: `m` < `n` - sort includes. Also, while you're at it, add an empty line after the local includes block.

File: auto_tests/crypto_test.c [hunk 4]
```diff
@@ -323,6 +324,86 @@ START_TEST(test_increment_nonce)
...
+    rand_bytes(src, sizeof(src));
+
+    uint8_t same[sizeof(src)];
+    memcpy(same, src, sizeof(src));
+
+    uint8_t not_same[sizeof(src)];
+    rand_bytes(not_same, sizeof(not_same));
+
+    clock_t same_mediane = memcmp_median(src, same, sizeof(src));
+    clock_t not_same_mediane = memcmp_median(src, not_same, sizeof(src));
```
Comment (iphydf) [inline auto_tests/crypto_test.c]: `median`, not `mediane`.

**Comment by iphydf** [comment]
Reviewed 8 of 9 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[auto_tests/crypto_test.c, line 342 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-K_Bx9TWdUIE6kr75SRo:-K_Bx9TXkE15ffdG-YDQ:bmgpwzx) ([raw file](https://github.com/toktok/c-toxcore/blob/ce6334c413d2101c841b096ffdfbc43e9d52a05b/auto_tests/crypto_test.c#L342)):*
> ```C
> 
> #define CRYPTO_TEST_MEMCMP_SIZE 1024*1024
> #define CRYPTO_TEST_MEMCMP_COUNT 10
> ```

I suggest making this number a lot larger (by 1 or 2 orders of magnitude). If the test takes too long then, reduce the `MEMCMP_SIZE` number. Increasing the count makes the result more precise and less likely to flake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/354)*

**Comment by Diadlo** [comment]
Review status: 4 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 211 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-KZmsyTQM_gx4vf3fxsg-r3-211:-K_C-y_3FnzBkN7k3Cmj:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/c180c7848e5faa67f0494712c36e9fb30ca4ce70/CMakeLists.txt#L211)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Lowercase function calls: `check_function_exists`.
</blockquote></details>

Done.

---

*[auto_tests/crypto_test.c, line 6 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-KZsTc_g9AsQaRX5v3Qd-r3-6:-K_C-zH-VO7GA-VfVITz:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/c180c7848e5faa67f0494712c36e9fb30ca4ce70/auto_tests/crypto_test.c#L6)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

`m` < `n` - sort includes. Also, while you're at it, add an empty line after the local includes block.
</blockquote></details>

Done.

---

*[auto_tests/crypto_test.c, line 392 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-KZsTc_g9AsQaRX5v3Qd-r3-392:-K_C0EGcPIL49hy21dCM:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/c180c7848e5faa67f0494712c36e9fb30ca4ce70/auto_tests/crypto_test.c#L392)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

`median`, not `mediane`.
</blockquote></details>

Done.

---

*[auto_tests/crypto_test.c, line 342 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-K_Bx9TWdUIE6kr75SRo:-K_C00LXb3I1TGzFdN-4:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/ce6334c413d2101c841b096ffdfbc43e9d52a05b/auto_tests/crypto_test.c#L342)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

I suggest making this number a lot larger (by 1 or 2 orders of magnitude). If the test takes too long then, reduce the `MEMCMP_SIZE` number. Increasing the count makes the result more precise and less likely to flake.
</blockquote></details>

Done.

---

*[CMakeLists.txt.user, line 1 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-KZuDNM-EZCQSXX5IBUf-r2-1:-K_C0FMwOtPxUdUjTkKL:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/8b71034fa256385ed573c4e618c36b095fca262b/CMakeLists.txt.user#L1)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

you might want to add `.user` to your gitignore
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/354)*

**Comment by iphydf** [comment]
Reviewed 1 of 2 files at r5.
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[auto_tests/crypto_test.c, line 346 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-K_C1Ta7n7qxuLaabGZ1:-K_C1Ta7n7qxuLaabGZ2:b29ot02) ([raw file](https://github.com/toktok/c-toxcore/blob/2551ad2bfe521f8b769192322fd6e5a868de1f38/auto_tests/crypto_test.c#L346)):*
> ```C
> #define CRYPTO_TEST_MEMCMP_EPS 10
> 
> make_quick_sort(clock_t)
> ```

Prepend `static` to `make_quick_sort` and all other file-local functions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/354)*

**Comment by Diadlo** [comment]
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[auto_tests/crypto_test.c, line 346 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-K_C1Ta7n7qxuLaabGZ1:-K_C2A1LVEPpGtsdFsa7:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/2551ad2bfe521f8b769192322fd6e5a868de1f38/auto_tests/crypto_test.c#L346)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Prepend `static` to `make_quick_sort` and all other file-local functions.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/354)*

**Comment by Diadlo** [comment]
cc @nurupo

**Comment by nurupo** [comment]
Reviewed 6 of 9 files at r3, 1 of 2 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[toxcore/crypto_core.c, line 40 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-K_aEsCJlETD3JZ9OH69:-K_aEsCJlETD3JZ9OH6A:bnqbfi7) ([raw file](https://github.com/toktok/c-toxcore/blob/d0939b55860b51c92ce0cf00187345e4d28084ec/toxcore/crypto_core.c#L40)):*
> ```C
> #include <sodium.h>
> #else
> #include "sodium_utils.h"
> ```

Shouldn't that be `#include "crypto_core_mem.c"`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/354)*

**Comment by Diadlo** [comment]
Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[toxcore/crypto_core.c, line 40 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-K_aEsCJlETD3JZ9OH69:-K_aGkNgiwIlJOPK0fTm:bwutas9) ([raw file](https://github.com/toktok/c-toxcore/blob/d0939b55860b51c92ce0cf00187345e4d28084ec/toxcore/crypto_core.c#L40)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Shouldn't that be `#include "crypto_core_mem.c"`?
</blockquote></details>

Line removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/354)*

**Comment by nurupo** [comment]
:lgtm_strong:

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/354#-:-K_aJlb7nlQEGQm2bMYO:b-jbsf8z)*

**Pull request merged**